File: python/n_sum.py
import argparse
import time

def n_sum(n: int, nums: list[int], target: int, left=0, right = None):
    if right is None:
        right = len(nums) - 1
    res = []

    # Base case
    if n == 2:
        while left < right:
            sum = nums[left] + nums[right]
            if sum > target:
                right -= 1
            elif sum < target:
                left += 1
            else:
                res.append([nums[left], nums[right]])
                left += 1
                right -= 1
                # skip duplicate answers
                while left < right and nums[left] == nums[left -1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
        return res
    # Recursive case
    else:
        for i in range(left, right + 1):
            if i > left and nums[i] == nums[i - 1]:
                continue
            sub_results = n_sum(n-1, nums, target - nums[i], i + 1, right)
            for num in sub_results:
                res.append([nums[i]] + num)
        return res

if __name__ == "__main__":
    # Parse args
    parser = argparse.ArgumentParser(description="Find all unique nSum combinations.")
    parser.add_argument("-n", "--n", type=int, required=True, help="How many numbers to sum (e.g., 4)")
    parser.add_argument("-t", "--target", type=int, required=True, help="The target sum")
    parser.add_argument("--nums", type=int, nargs="+", required=True, help="The list of numbers separated by spaces")
    args = parser.parse_args()

    # Start timer
    start_time = time.perf_counter()

    # Run the algo
    args.nums.sort()
    result = n_sum(args.n, args.nums, args.target)

    # Stop the stopwatch
    end_time = time.perf_counter()

    # Print the results
    print(f"The {len(result)} unique combinations of {args.n} numbers that add to {args.target} are:")
    for res in result:
        print(f"> {res}")

    print(f"\n Time taken: {end_time - start_time:.4f} seconds")

File: cpp/n_sum.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <vector>

void n_sum(int n, std::vector<int> &nums, int target,
           std::vector<std::vector<int>> &res, std::vector<int> &current_path,
           int left = 0, int right = -1) {
  if (right == -1) {
    right = nums.size() - 1;
  }

  // Base case
  if (n == 2) {
    while (left < right) {
      long long sum = (long long)nums[left] + nums[right];
      if (sum > target) {
        right -= 1;
      } else if (sum < target) {
        left += 1;
      } else {

        // Match found
        current_path.push_back(nums[left]);
        current_path.push_back(nums[right]);

        res.push_back(current_path);

        current_path.pop_back();
        current_path.pop_back();

        left += 1;
        right -= 1;
        while (left < right and nums[left] == nums[left - 1]) {
          left += 1;
        }
        while (left < right and nums[right] == nums[right + 1]) {
          right -= 1;
        }
      }
    }
  } else {
    // Recursive case
    for (int i = left; i <= right; i++) {
      if (i > left and nums[i] == nums[i - 1]) {
        continue;
      }

      // Pruning
      if (right - i + 1 < n)
        break;
      long long min_sum = 0, max_sum = 0;
      for (int k = 0; k < n; ++k) {
        min_sum += nums[i + k];
        max_sum += nums[right - k];
      }
      if (target < min_sum)
        break;
      if (target > max_sum)
        continue;

      current_path.push_back(nums[i]);

      n_sum(n - 1, nums, target - nums[i], res, current_path, i + 1, right);

      current_path.pop_back();
    }
  }
}

int main() {
  int n, target;
  std::cout << "Enter n: ";
  std::cin >> n;

  std::cout << "Enter target: ";
  std::cin >> target;

  std::cout << "Enter numbers (seperated by spaces) end input with any letter:\n";
  std::vector<int> nums;
  int num;

  while (std::cin >> num) {
    nums.push_back(num);
  }

  // Start the timer
  auto start = std::chrono::high_resolution_clock::now();

  // Run the algo
  std::sort(nums.begin(), nums.end());

  std::vector<std::vector<int>> result;
  std::vector<int> current_path;

  n_sum(n, nums, target, result, current_path);

  // Stop the timer
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> duration = end - start;

  std::cout << "The " << result.size() << " unique combinations are:\n";
  for (const auto &combo : result) {
    std::cout << "> ";
    for (int num : combo) {
      std::cout << num << " ";
    }
    std::cout << "\n";
  }

  std::cout << "\n Time taken: " << duration.count() << " seconds\n";

  return 0;
}
